core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -589,26 +589,39 @@ class Partition(val topicPartition: TopicPartition,
       case Some(followerReplica) =>
         // No need to calculate low watermark if there is no delayed DeleteRecordsRequest
         val oldLeaderLW = if (delayedOperations.numDelayedDelete > 0) lowWatermarkIfLeader else -1L
+        val prevFollowerEndOffset = followerReplica.logEndOffset
         followerReplica.updateFetchState(
           followerFetchOffsetMetadata,
           followerStartOffset,
           followerFetchTimeMs,
           leaderEndOffset)
+
+        // when updateFetchState does not result in any changes, we do not need to
+        // check for ISR expansion or delayed request completions
         val newLeaderLW = if (delayedOperations.numDelayedDelete > 0) lowWatermarkIfLeader else -1L
         // check if the LW of the partition has incremented
         // since the replica's logStartOffset may have incremented
         val leaderLWIncremented = newLeaderLW > oldLeaderLW
+
         // check if we need to expand ISR to include this replica
         // if it is not in the ISR yet
-        val followerFetchOffset = followerFetchOffsetMetadata.messageOffset
-        val leaderHWIncremented = maybeExpandIsr(followerReplica, followerFetchTimeMs)
+        if (!inSyncReplicaIds(followerId))
+          maybeExpandIsr(followerReplica, followerFetchTimeMs)
+
+        // check if the HW of the partition can now be incremented
+        // since the replica may already be in the ISR and its LEO has just incremented
+        val leaderHWIncremented = if (prevFollowerEndOffset != followerReplica.logEndOffset) {
+          leaderLogIfLocal.exists(leaderLog => maybeIncrementLeaderHW(leaderLog, followerFetchTimeMs))
```

Comment:
We were previously incrementing the hw atomically with the isr expansion. Have we verified that it's ok to change that?

Reply:
Hmm, I will look into that. The second problem is that we should probably be taking out a lock in maybeIncrementLeaderHW even if the update does not need to be atomic.

Reply:
Yeah, I was thinking about this too. We probably need to grab the read side of the `leaderIsrUpdateLock` in `maybeIncrementLeaderHw` to protect from a concurrent ISR change. The second thing is that `Log.maybeIncrementHighWatermark` probably needs to be holding the lock when checking for monotonicity. Currently it only grabs the lock when actually updating the value.

Reply:
@hachikuji the changes you suggested seem good to me. Let me know what you think and whether you think I've implemented the locking at the right granularity.